src/proto/h1/conn.rs

```diff
@@ -493,22 +485,18 @@ where
     }
 
     pub(crate) fn can_write_head(&self) -> bool {
-        if !T::should_read_first() {
-            if let Reading::Closed = self.state.reading {
-                return false;
-            }
+        if !T::should_read_first() && matches!(self.state.reading, Reading::Closed) {
+            return false;
         }
+
         match self.state.writing {
             Writing::Init => self.io.can_headers_buf(),
             _ => false,
         }
     }
 
     pub(crate) fn can_write_body(&self) -> bool {
-        match self.state.writing {
-            Writing::Body(..) => true,
-            Writing::Init | Writing::KeepAlive | Writing::Closed => false,
-        }
+        matches!(self.state.writing, Writing::Body(..))
```

Comment:
In some of these cases, the exhaustive match was on purpose to make sure any changes to the variants were always considered here... I do like that this makes it more succinct. I don't know which is the better choice...

Reply:
I totally understand where you're coming from here. State machines are really tricky to get right.

Which specific cases are you referring to aside from the one highlighted here? I'd be glad to revert them just to relieve any doubts for future maintainability. 👍

Reply:
Probably just the ones that were already exhaustive (didn't have a `_` branch).